### PR TITLE
DDCNL-3571 CySP COPE Update NISP Frontend

### PR DESCRIPTION
**Pull request by gianick**

### CODE REVIEW

File: app/uk/gov/hmrc/nisp/errorHandler/ErrorHandler.scala [hunk 1]
```diff
@@ -18,13 +18,19 @@ package uk.gov.hmrc.nisp.errorHandler
 
 import com.google.inject.Inject
 import play.api.i18n.MessagesApi
-import play.api.mvc.Request
+import play.api.mvc.{Request, RequestHeader, Result}
```
Comment (kurtlogan) [inline app/uk/gov/hmrc/nisp/errorHandler/ErrorHandler.scala:21]: Some imports here aren't used

File: app/uk/gov/hmrc/nisp/models/StatePensionExclusionFiltered.scala [hunk 2]
```diff
@@ -17,19 +17,44 @@
...
+case class StatePensionExclusionFiltered(
+                                 exclusion: Exclusion,
+                                 pensionAge: Option[Int] = None,
+                                 pensionDate: Option[LocalDate] = None,
+                                 statePensionAgeUnderConsideration: Option[Boolean] = None
+                              ) extends StatePensionExclusionFilter
+
+case class StatePensionExclusionFilteredWithCopeDate(
+                                                      exclusion: Exclusion,
+                                                      pensionAge: Option[Int] = None,
```
Comment (kurtlogan) [inline app/uk/gov/hmrc/nisp/models/StatePensionExclusionFiltered.scala:48]: Is pensionAge, pensionDate and statePensionAgeUnderConsideration needed for this scenario?

File: app/uk/gov/hmrc/nisp/services/StatePensionService.scala [hunk 3]
```diff
@@ -53,6 +55,11 @@ class StatePensionService @Inject()(statePensionConnector: StatePensionConnector
           Left(StatePensionExclusionFiltered(Exclusion.Dead))
         case ex: Upstream4xxResponse if ex.upstreamResponseCode == FORBIDDEN && ex.message.contains(exclusionCodeManualCorrespondence) =>
           Left(StatePensionExclusionFiltered(Exclusion.ManualCorrespondenceIndicator))
+        case ex: Upstream4xxResponse if ex.upstreamResponseCode == FORBIDDEN && ex.message.contains(exclusionCodeCopeProcessingFailed) =>
+          Left(StatePensionExclusionFiltered(Exclusion.CopeProcessingFailed))
+        case ex: Upstream4xxResponse if ex.upstreamResponseCode == FORBIDDEN && ex.message.contains(exclusionCodeCopeProcessing) =>
+          Left(StatePensionExclusionFilteredWithCopeDate(Exclusion.CopeProcessing, copeAvailableDate = getDateWithRegex(ex.message)))
+        // Case match not exhaustive
```
Comment (kurtlogan) [inline app/uk/gov/hmrc/nisp/services/StatePensionService.scala:62]: It's fine that this case match isn't exhaustive as any other exceptions will be propagated up which is what we want

File: app/uk/gov/hmrc/nisp/services/StatePensionService.scala [hunk 4]
```diff
@@ -77,4 +84,13 @@ class StatePensionService @Inject()(statePensionConnector: StatePensionConnector
       throw new RuntimeException(s"Un-accounted for exclusion in NispConnectionNI: $exclusions")
     }
   }
+
+  private def getDateWithRegex(copeResponse: String): LocalDate = {
+    val copeResponseDateCapturingRegex: Regex = """(?:.*)(?:'\{"errorCode":"EXCLUSION_COPE_PROCESSING","copeDataAvailableDate":\")(\d{4}-\d{2}-\d{2})(?:\"}')""".r
```
Comment (kurtlogan) [inline app/uk/gov/hmrc/nisp/services/StatePensionService.scala:89]: This is a little brittle, I would consider having a case class that represents the repsonse and parse it that way then extract the date

File: conf/messages.cy [hunk 5]
```diff
@@ -167,6 +167,14 @@ nisp.excluded.haveReached = Rydych wedi cyrraedd oedran Pensiwn y Wladwriaeth ar
 nisp.excluded.willReach = Byddwch yn cyrraedd oedran Pensiwn y Wladwriaeth ar <span class="nowrap">{0}</span>
 nisp.nirecord.summary.youhave = Mae gennych:
 
+nisp.excluded.cope.processing.h1 = Your State Pension Forecast
```
Comment (kurtlogan) [inline conf/messages.cy:170]: Do we have the welsh yet?

File: conf/messages [hunk 6]
```diff
@@ -170,6 +170,14 @@ nisp.excluded.haveReached = You reached State Pension age on {0}
 nisp.excluded.willReach = You’ll reach State Pension age on <span class="nowrap">{0}</span>
 nisp.nirecord.summary.youhave = You have:
 
+nisp.excluded.cope.processing.h1 = Your State Pension Forecast
```
Comment (kurtlogan) [inline conf/messages:173]: This content doesn't look to match what is in the ticket

File: app/uk/gov/hmrc/nisp/views/excluded_cope_failed.scala.html [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    showUrBannerOn == PostStatePensionAge | showUrBannerOn == IsleOfMan
+}
+
+@unauthenticatedMain(MainTemplateParams(userLoggedIn = true, browserTitle = Some(messages("nisp.main.title")),
+            pageTitle = Some(messages("nisp.excluded.cope.failed.h1")), articleClasses = Some("mainpage"))
+) {
+    <div>
+        <p>@Html(messages("nisp.excluded.cope.processing.failed"))</p>
+
+        <a id="futurePensionLink" href="@appConfig.futurePensionUrl">@Html(messages("nisp.excluded.cope.processing.failed.link"))</a>
```
Comment (kurtlogan) [inline app/uk/gov/hmrc/nisp/views/excluded_cope_failed.scala.html:51]: There looks like a lot more content on the screenshot than what is here

File: app/uk/gov/hmrc/nisp/errorHandler/ErrorHandler.scala (hunk 1 above)
Comment (gianick) [reply app/uk/gov/hmrc/nisp/errorHandler/ErrorHandler.scala]: Thanks. I optimised the imports.

File: conf/messages.cy (hunk 5 above)
Comment (gianick) [reply conf/messages.cy]: I've added the Welsh translations.

File: conf/messages (hunk 6 above)
Comment (gianick) [reply conf/messages]: Updated the pages to match revised designs/content.

File: app/uk/gov/hmrc/nisp/views/excluded_cope_failed.scala.html (hunk 7 above)
Comment (gianick) [reply app/uk/gov/hmrc/nisp/views/excluded_cope_failed.scala.html]: Updated the pages to match revised designs/content.

File: app/uk/gov/hmrc/nisp/services/StatePensionService.scala (hunk 4 above)
Comment (gianick) [reply app/uk/gov/hmrc/nisp/services/StatePensionService.scala:89]: Ticket created to refactor this.

**Review by matthew-mcnally — APPROVED** [review]

**Pull request merged**